exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/diagnosiskey/DiagnosisKeyDao.java

```diff
@@ -118,24 +119,57 @@ public boolean verify(int verificationId, String requestChecksum, long totalKeyC
         }
     }
 
+    public List<TemporaryExposureKey> fetchAvailableKeysForEfgs(long operationId) {
+        LOG.info("Fetching queued keys not sent to efgs.");
+        String sql = "select key_data, rolling_period, rolling_start_interval_number, transmission_risk_level, " +
+                "visited_countries, days_since_onset_of_symptoms, origin, consent_to_share " +
+                "from en.diagnosis_key " +
+                "where efgs_operation = :efgs_operation " +
+                "order by key_data";
+
+        return new ArrayList<>(jdbcTemplate.query(sql, Map.of("efgs_operation", operationId), (rs, i) -> mapKey(rs)));
+    }
+
+
+
+    @Transactional
+    public void addInboundKeys(List<TemporaryExposureKey> keys, int interval) {
+        if (!keys.isEmpty()) {
+            boolean finished = false;
+            long operationId = federationOperationDao.startInboundOperation();
```

Comment:
I think it would be cleaner if these calls came from the service doing the sync, not this dao. That keeps the dao's less connected and makes it possible to implement "fetch only once" despite multiple nodes. Of course, it needs more logic for timeout and such as well - this method does a single transaction so there's no middle states.

Reply:
Yep, I'll leave that to next refactoring.

Reply:
Let's see how this ends. I'll keep this in mind when there is refactoring time.